Spark Python Kubernetes Kernel starts, but Spark is never initialized

## Description
Built Jupyter Enterprise Gateway and deployed on kubernetes. Regular kernels (R, Python, Scala) all work fine. The Spark kernels took a while to start, but then. However the Spark context is never initialized.

sc.version returns:

Py4JJavaError: An error occurred while calling None.org.apache.spark.api.java.JavaSparkContext.
: java.lang.NoClassDefFoundError: io/fabric8/kubernetes/client/KubernetesClient
	at org.apache.spark.scheduler.cluster.k8s.KubernetesClusterManager.createSchedulerBackend(KubernetesClusterManager.scala:64)
	at org.apache.spark.SparkContext$.org$apache$spark$SparkContext$$createTaskScheduler(SparkContext.scala:2788)
	at org.apache.spark.SparkContext.<init>(SparkContext.scala:493)
	at org.apache.spark.api.java.JavaSparkContext.<init>(JavaSparkContext.scala:58)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:247)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:238)
	at py4j.commands.ConstructorCommand.invokeConstructor(ConstructorCommand.java:80)
	at py4j.commands.ConstructorCommand.execute(ConstructorCommand.java:69)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: io.fabric8.kubernetes.client.KubernetesClient
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	... 15 more

## Environment

- Enterprise Gateway built from master branch
- Platform: Kubernetes
- Others: Jupyter 1.0.0, Jupyter Client 5.3.4, Jupyterlab 1.2.6, notebook 6.0.3, jupyterlab_server 1.0.6

--- a/etc/docker/enterprise-gateway/Dockerfile
+++ b/etc/docker/enterprise-gateway/Dockerfile
@@ -1,6 +1,6 @@
 FROM jupyter/minimal-notebook
 
-ENV SPARK_VER 2.4.1
+ENV SPARK_VER 2.4.5
 ENV SPARK_HOME /opt/spark
 ENV JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
 

--- a/etc/docker/kernel-image-puller/kernel_image_puller.py
+++ b/etc/docker/kernel-image-puller/kernel_image_puller.py
@@ -86,8 +86,19 @@ def pull_image(image_name):
     """
     if policy == POLICY_IF_NOT_PRESENT:
         if image_name in pulled_images:
-            logger.info("Image '{}' already pulled and policy is '{}'.".format(image_name, policy))
-            return
+            # Image has been pulled, but make sure it still exists.  If it doesn't exist
+            # let this drop through to actual pull
+            logger.info("Image '{}' already pulled and policy is '{}'.  Checking existence.".
+                        format(image_name, policy))
+            try:
+                t1 = time.time()
+                docker_client.images.get(image_name)
+                t2 = time.time()
+                logger.debug("Checked existence of image '{}' in {:.3f} secs.".format(image_name, t2 - t1))
+                return
+            except NotFound:
+                pulled_images.remove(image_name)
+                logger.warning("Previously pulled image '{}' was not found - attempting pull...".format(image_name))
 
     logger.debug("Pulling image '{}'...".format(image_name))
     try:

--- a/etc/docker/kernel-scala/Dockerfile
+++ b/etc/docker/kernel-scala/Dockerfile
@@ -2,7 +2,7 @@ FROM elyra/spark:v2.4.5
 
 ADD jupyter_enterprise_gateway_kernel_image_files*.tar.gz /usr/local/bin/
 
-# Craete/setup the jovyan system user
+# Create/setup the jovyan system user
 RUN adduser --system -uid 1000 jovyan --ingroup users && \
     chown jovyan:users /usr/local/bin/bootstrap-kernel.sh && \
 	chmod 0755 /usr/local/bin/bootstrap-kernel.sh && \

--- a/etc/docker/kernel-scala/Dockerfile
+++ b/etc/docker/kernel-scala/Dockerfile
@@ -1,18 +1,9 @@
-FROM elyra/spark:v2.4.1
-
-# Depending on the version of Kubernetes, some Spark jobs get
-# failures attempting to create executors. These steps update
-# the kubernetes client to 4.4.2 and will not be necessary
-# once Spark 2.4.5 or 3.0 have been released.
-# See https://issues.apache.org/jira/browse/SPARK-28921
-RUN rm -f $SPARK_HOME/jars/kubernetes-*.jar
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-client/4.4.2/kubernetes-client-4.4.2.jar $SPARK_HOME/jars
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-model/4.4.2/kubernetes-model-4.4.2.jar $SPARK_HOME/jars
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-model-common/4.4.2/kubernetes-model-common-4.4.2.jar $SPARK_HOME/jars
+FROM elyra/spark:v2.4.5
 
 ADD jupyter_enterprise_gateway_kernel_image_files*.tar.gz /usr/local/bin/
 
-RUN adduser -S -u 1000 -G users jovyan && \
+# Craete/setup the jovyan system user
+RUN adduser --system -uid 1000 jovyan --ingroup users && \
     chown jovyan:users /usr/local/bin/bootstrap-kernel.sh && \
 	chmod 0755 /usr/local/bin/bootstrap-kernel.sh && \
 	chmod 0777 /opt/spark/work-dir && \

--- a/etc/docker/kernel-spark-py/Dockerfile
+++ b/etc/docker/kernel-spark-py/Dockerfile
@@ -4,7 +4,7 @@ ARG TAG
 # Ubuntu 18.04.1 LTS Bionic
 FROM $HUB_ORG/kernel-py:$TAG
 
-ENV SPARK_VER 2.4.1
+ENV SPARK_VER 2.4.5
 ENV SPARK_HOME /opt/spark
 ENV KERNEL_LANGUAGE python
 ENV R_LIBS_USER $R_LIBS_USER:${SPARK_HOME}/R/lib
@@ -33,17 +33,7 @@ RUN cd /opt/ && \
     wget https://raw.githubusercontent.com/apache/spark/v${SPARK_VER}/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh && \
     chmod a+x /opt/entrypoint.sh && \
     sed -i 's/tini -s/tini -g/g' /opt/entrypoint.sh && \
-    ln -sfn /opt/conda/bin/tini /sbin/tini
-
-# Depending on the version of Kubernetes, some Spark jobs get
-# failures attempting to create executors. These steps update
-# the kubernetes client to 4.4.2 and will not be necessary
-# once Spark 2.4.5 or 3.0 have been released.
-# See https://issues.apache.org/jira/browse/SPARK-28921
-RUN rm -f $SPARK_HOME/jars/kubernetes-*.jar
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-client/4.4.2/kubernetes-client-4.4.2.jar $SPARK_HOME/jars
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-model/4.4.2/kubernetes-model-4.4.2.jar $SPARK_HOME/jars
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-model-common/4.4.2/kubernetes-model-common-4.4.2.jar $SPARK_HOME/jars
+    ln -sfn /opt/conda/bin/tini /usr/bin/tini
 
 WORKDIR $SPARK_HOME/work-dir
 # Ensure that work-dir is writable by everyone

--- a/etc/docker/kernel-spark-r/Dockerfile
+++ b/etc/docker/kernel-spark-r/Dockerfile
@@ -5,7 +5,7 @@ FROM $HUB_ORG/kernel-r:$TAG
 
 USER root
 
-ENV SPARK_VER 2.4.1
+ENV SPARK_VER 2.4.5
 ENV SPARK_HOME /opt/spark
 ENV KERNEL_LANGUAGE=R
 ENV R_LIBS_USER $R_LIBS_USER:${R_HOME}/library:${SPARK_HOME}/R/lib
@@ -29,17 +29,7 @@ RUN cd /opt/ && \
     wget https://raw.githubusercontent.com/apache/spark/v${SPARK_VER}/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh && \
     chmod a+x /opt/entrypoint.sh && \
     sed -i 's/tini -s/tini -g/g' /opt/entrypoint.sh && \
-    ln -sfn /opt/conda/bin/tini /sbin/tini
-
-# Depending on the version of Kubernetes, some Spark jobs get
-# failures attempting to create executors. These steps update
-# the kubernetes client to 4.4.2 and will not be necessary
-# once Spark 2.4.5 or 3.0 have been released.
-# See https://issues.apache.org/jira/browse/SPARK-28921
-RUN rm -f $SPARK_HOME/jars/kubernetes-*.jar
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-client/4.4.2/kubernetes-client-4.4.2.jar $SPARK_HOME/jars
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-model/4.4.2/kubernetes-model-4.4.2.jar $SPARK_HOME/jars
-ADD https://repo1.maven.org/maven2/io/fabric8/kubernetes-model-common/4.4.2/kubernetes-model-common-4.4.2.jar $SPARK_HOME/jars
+    ln -sfn /opt/conda/bin/tini /usr/bin/tini
 
 WORKDIR $SPARK_HOME/work-dir
 # Ensure that work-dir is writable by everyone

